addons/stock/report/product_templates.xml

```diff
@@ -38,17 +38,28 @@
                 </t>
             </t>
         </template>
+
         <template id="label_lot_template_view">
             <t t-foreach="docs" t-as="lot">
                 <t t-translation="off">
 ^XA
 ^FO100,50
-^A0N,44,33^FD<t t-esc="lot.product_id.display_name"/>^FS
+^A0N,44,33^FD<t t-out="lot.product_id.display_name"/>^FS
 ^FO100,100
-^A0N,44,33^FDLN/SN: <t t-esc="lot.name"/>^FS
+^A0N,44,33^FDLN/SN: <t t-out="lot.name"/>^FS
+                    <t t-if="env.user.has_group('stock.group_stock_lot_print_gs1') and env['ir.actions.report'].get_datamatrix_state() != 'broken'">
+                        <t t-if="lot.product_id.valid_ean" t-set="final_barcode" t-value="'01' + '0' * (14 - len(lot.product_id.barcode)) + lot.product_id.barcode"/>
+                        <t t-if="lot.product_id.tracking == 'lot'" t-set="final_barcode" t-value="(final_barcode or '') + '10' + '0' * (20 - len(lot.name)) + lot.name"/>
+                        <t t-elif="lot.product_id.tracking == 'serial'" name="datamatrix_sn" t-set="final_barcode" t-value="(final_barcode or '') + '21' + '0' * (20 - len(lot.name)) + lot.name"/>

```

Comment:
I guess the `'0' * (20 - len(lot.name))` is done to avoid to put the FNC1 into the barcode ?
The problem with this solution is the serial/lot number isn't the same as `'lot-0001'` =/= `'000000000000lot-0001'`.

Padding with 0 for the product's barcode works because the product's barcode have to be 14 digits, so if your product's barcode is `'12345670'` , we will check for `'0100000012345670'`, which is not the case for variable length AI parts.

I suggest to always place the lot/serial part of the barcode at the end so we don't have to end it by the FNC1 or fix its length, but it is a little tricky with the override...